cds-snc/platform-forms-client
# refactor: update confirm api to a server action

# Summary | Résumé

This PR refactors the confirm API to a server action as well implements database throttling to handle larger read / write loads.

Instead of trying to implement exponential back off and retries manually this PR instead uses the built in retry services that are included by default in the AWS SDK.  The number of retries for the dynamodb service has been increased from 3 to 15.  In LocalStack with the dynamodb set to return throttle capacity exceptions 50% of the time we can see the AWS retry logic at work.

As an additional throttling mechanism a 200ms delay is introduced between batched calls to more evenly spread out the load over the partition key (form ID)

| BatchGetCommand | TransactWriteItems Command | Query Command | 
| ---------- | ------------- | ------------- |
| ![Screenshot 2024-07-19 at 8 06 45 AM](https://github.com/user-attachments/assets/4c3abd04-8195-41b9-9764-760712da2f13) | ![Screenshot 2024-07-19 at 7 55 50 AM](https://github.com/user-attachments/assets/3279d875-82da-48df-a92c-bb3013dec4d8) | ![Screenshot 2024-07-18 at 12 32 50 PM](https://github.com/user-attachments/assets/c377138c-c29b-40e9-b95e-3ab4ff669f87) |

# Test instructions | Instructions pour tester la modification

Setup:
- Publish a Form
- Generate a ton, and I mean of ton of responses for it
- Set the response download limit to something ridiculous like 1000

Test
- Download 1000 responses at once
- Confirm 1000 responses at once

The actions should not throw errors and should complete, albeit a little more slowly, for a user.
If there are dynamodb warnings thrown in Slack it should have no impact on the user.

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

Still need to refactor "report a problem" and "unprocessed submissions" APIs to server actions.

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:
